### PR TITLE
Improve inline documentation of config defaults and use it to generate website counterpart

### PR DESCRIPTION
**Pull request by astorije**
This is a first part, pre-requisite for https://github.com/thelounge/thelounge.github.io/pull/84.

So the way it works is that the script generates a Markdown from all the inline comments and puts it in [`_includes/config.js.md` of the doc folder](https://github.com/thelounge/thelounge.github.io/pull/84/files#diff-df923e880eb889dae72e108ad9bb7165). The generated file needs to be checked in and it is then read [by `_docs/configuration.md`](https://github.com/thelounge/thelounge.github.io/pull/84/files#diff-5b88d6756daca0ac2a4e130b4121f431R19).
When the website gets built, it uses the same content than the actual config file to reduce [this kind of totally out-of-sync mess](https://thelounge.chat/docs/server/configuration.html).
It then looks like this: [configuration.html](https://deploy-preview-84--thelounge-chat.netlify.com/docs/configuration.html). Note that my PR is not up-to-date with the latest defaults as we speak. Also I think this page needs needs a TOC but that would affect the website/docs layout and not this file/PR.

I know this isn't perfect, but hey, it's a first step :) We can do a lot of things to improve keeping them in sync, but my goal was to keep the doc both in the config file _and_ the website while trying to keep them in sync more easily than it is at the moment.

I have checked that my rewrite of `defaults/config.js` is 1:1 in actual code with `master`. Keys are in a different order, but all the keys are there and the values are the same.

### CODE REVIEW

File: defaults/config.js [hunk 1]
```diff
@@ -1,510 +1,428 @@
...
 	//
-	// @type     string
-	// @default  "The Lounge - https://thelounge.chat"
+	// This value is set to `["sqlite", "text"]` by default.
+	messageStorage: ["sqlite", "text"],
+
+	// ### `useHexIp`
 	//
-	leaveMessage: "The Lounge - https://thelounge.chat",
+	// When set to `true`, users' IP addresses will encoded has hex.
```
Comment (xPaw) [inline defaults/config.js]: `has` -> `as`

And specify that this is encodes it in the `username` field and it only supports IPv4.

File: defaults/config.js [hunk 2]
```diff
@@ -1,510 +1,428 @@
...
+	// This value is set to `10000` by default.
+	maxHistory: 10000,
 
+	// ### `https`
 	//
-	// Prefetch URLs
+	// These three settings are used to run The Lounge using encrypted HTTP/2 on
+	// the server side. This will fallback to regular HTTPS if HTTP/2 is not
+	// supported.
 	//
```
Comment (xPaw) [inline defaults/config.js]: Add a line saying reverse proxy is preferred if you want more options / live reload of certs

File: scripts/generate-config-doc.js [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+	const infoBlock = `<!--
+DO NOT EDIT THIS FILE MANUALLY.
+Content for the following is generated by this script in the main repo:
+https://github.com/thelounge/lounge/blob/master/scripts/generate-config-doc.js
+Last updated at ${Date()} by ${username}
+-->`;
+
+	const generatedContent = `${infoBlock}\n\n${extractedDoc}\n${infoBlock}\n`;
+
+	fs.writeFileSync(docPath, generatedContent);
```
Comment (xPaw) [inline scripts/generate-config-doc.js]: You wrap in async but use sync version here

File: defaults/config.js [hunk 4]
```diff
@@ -1,510 +1,428 @@
...
+	// The available keys for the `defaults` object are:
+	//
+	// - `name`: Name to display in the channel list of The Lounge. This value is
+	//   not forwarded to the IRC network.
+	// - `host`
+	// - `port`: Usually 6667 for unencrypted connections and 6697 for
+	//   connections encrypted with TLS.
+	// - `password`
+	// - `tls`: Enable TLS connections
+	// - `rejectUnauthorized`: Whether or not the server certificate should be
```
Comment (xPaw) [inline defaults/config.js]: `or not` is not necessary

File: defaults/config.js [hunk 5]
```diff
@@ -1,510 +1,428 @@
...
+	//   verified against the list of supplied CAs by your Node.js installation.
```
Comment (xPaw) [inline defaults/config.js]: Expand CAs to certificate authorities.

File: defaults/config.js [hunk 6]
```diff
@@ -1,510 +1,428 @@
...
 	//
 	// The authentication process works as follows:
 	//
-	//   1. Lounge connects to the LDAP server with its system credentials
-	//   2. It performs a LDAP search query to find the full DN associated to the
-	//      user requesting to log in.
-	//   3. Lounge tries to connect a second time, but this time using the user's
-	//      DN and password. Auth is validated iff this connection is successful.
+	// 1. The Lounge connects to the LDAP server with its system credentials.
+	// 2. It performs a LDAP search query to find the full DN associated to the
```
Comment (xPaw) [inline defaults/config.js]: `an LDAP`?

File: defaults/config.js (hunk 1 above)
Comment (astorije) [reply defaults/config.js]: Done, let me know what you think.

File: defaults/config.js [hunk 7]
```diff
@@ -1,510 +1,428 @@
...
-	lockNetwork: false,
+	// ### `prefetchStorage`
 
+	// When set to `true`, The Lounge will store and proxy prefetched images and
+	// thumbnails on the filesystem rather than directly display the content at
+	// the original URLs.
 	//
-	// Hex IP
+	// This improves security and privacy by not exposing the client IP address,
+	// and always loading images from The Lounge and making all assets secure,
```
[the remark this answers is not in the document]
Comment (astorije) [reply defaults/config.js]: > This improves security and privacy by [...] always loading images from The Lounge and making all assets secure [...]

So `loading` is intended.

File: defaults/config.js (hunk 2 above)
Comment (astorije) [reply defaults/config.js]: Will do.

File: scripts/generate-config-doc.js (hunk 3 above)
Comment (astorije) [reply scripts/generate-config-doc.js]: Yeah, I forget why but I used a promisified function above, and I couldn't be bothered to make everything beautifully async.
Async brings nothing here in terms of perf or whatnot, but I'll make the script consistent one way or the other.

File: defaults/config.js [hunk 8]
```diff
@@ -1,510 +1,428 @@
...
-	// text: Text file per network/channel in user folder
-	// sqlite: Messages are stored in SQLite, this allows them to be reloaded on server restart
+	// Set `socket.io` transports.
 	//
-	// @type     array
-	// @default  ["sqlite", "text"]
-	//
-	messageStorage: ["sqlite", "text"],
+	// This value is set to `["polling", "websocket"]` by default.
+	transports: ["polling", "websocket"],
```
[the remark this answers is not in the document]
Comment (xPaw) [reply defaults/config.js]: It was a generic comment yeah, we shouldn't do this in this PR.

They mention making websocket default for socket.io 3.0 here https://github.com/socketio/socket.io/issues/3250

[the remark this answers is not in the document]
Comment (astorije) [reply defaults/config.js]: Maybe open an issue to track this? (or maybe we already have one?)

File: scripts/generate-config-doc.js (hunk 3 above)
Comment (astorije) [reply scripts/generate-config-doc.js]: Done, all sync now. Clearer indeed. I'll finish addressing the other comment or 2 you left tomorrow.

File: defaults/config.js (hunk 2 above)
Comment (astorije) [reply defaults/config.js]: Done, linked to the [upcoming guide on reverse proxies](https://5acafe274ed62f0a99ee7c04--thelounge-chat.netlify.com/docs/guides/reverse-proxies.html).

**Comment by astorije** [comment]
Alright @xPaw, I addressed all your comments. Let me know if there are other things you want changed.
@McInkay or @MaxLeiter, want to take another look?

**Pull request merged**